File: java/shared/java/org/signal/libsignal/protocol/ecc/Curve.java
/**
 * Copyright (C) 2013-2022 Open Whisper Systems
 *
 * Licensed according to the LICENSE file in this repository.
 */
package org.signal.libsignal.protocol.ecc;
import org.signal.libsignal.protocol.InvalidKeyException;

public class Curve {
  public static final int CURVE_25519_TYPE   = 0x05;

  public static ECKeyPair generateKeyPair() {
    ECPrivateKey privateKey = ECPrivateKey.generate();
    ECPublicKey publicKey = privateKey.publicKey();
    return new ECKeyPair(publicKey, privateKey);
  }

  public static ECPublicKey decodePoint(byte[] bytes, int offset)
      throws InvalidKeyException
  {
    if (bytes == null || bytes.length - offset < 1) {
      throw new InvalidKeyException("No key type identifier");
    }

    return new ECPublicKey(bytes, offset);
  }

  public static ECPrivateKey decodePrivatePoint(byte[] bytes) {
    return new ECPrivateKey(bytes);
  }

  public static byte[] calculateAgreement(ECPublicKey publicKey, ECPrivateKey privateKey)
      throws InvalidKeyException
  {
    if (publicKey == null) {
      throw new InvalidKeyException("public value is null");
    }

    if (privateKey == null) {
      throw new InvalidKeyException("private value is null");
    }

    return privateKey.calculateAgreement(publicKey);
  }

  public static boolean verifySignature(ECPublicKey signingKey, byte[] message, byte[] signature)
      throws InvalidKeyException
  {
    if (signingKey == null || message == null || signature == null) {
      throw new InvalidKeyException("Values must not be null");
    }

    return signingKey.verifySignature(message, signature);
  }

  public static byte[] calculateSignature(ECPrivateKey signingKey, byte[] message)
      throws InvalidKeyException
  {
    if (signingKey == null || message == null) {
      throw new InvalidKeyException("Values must not be null");
    }

    return signingKey.calculateSignature(message);
  }

}

File: java/shared/java/org/signal/libsignal/protocol/ecc/Curve.kt
//
// Copyright 2013-2016 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

@file:JvmName("Curve")

package org.signal.libsignal.protocol.ecc

import org.signal.libsignal.protocol.InvalidKeyException

class Curve {
  companion object {
    const val DJB_TYPE = 0x05

    @JvmStatic
    fun generateKeyPair(): ECKeyPair {
      val privateKey = ECPrivateKey.generate()
      val publicKey = privateKey.publicKey()
      return ECKeyPair(publicKey, privateKey)
    }

    @JvmStatic
    @Throws(InvalidKeyException::class)
    fun decodePoint(bytes: ByteArray?, offset: Int): ECPublicKey {
      if (bytes == null || bytes.size - offset < 1) {
        throw InvalidKeyException("No key type identifier")
      }

      return ECPublicKey(bytes, offset)
    }

    @JvmStatic
    @Throws(InvalidKeyException::class)
    fun decodePrivatePoint(bytes: ByteArray?): ECPrivateKey {
      return ECPrivateKey(bytes)
    }

    @JvmStatic
    @Throws(InvalidKeyException::class)
    fun calculateAgreement(publicKey: ECPublicKey?, privateKey: ECPrivateKey?): ByteArray {
      if (publicKey == null) {
        throw InvalidKeyException("public value is null")
      }

      if (privateKey == null) {
        throw InvalidKeyException("private value is null")
      }

      return privateKey.calculateAgreement(publicKey)
    }

    @JvmStatic
    @Throws(InvalidKeyException::class)
    fun verifySignature(signingKey: ECPublicKey?, message: ByteArray?, signature: ByteArray?): Boolean {
      if (signingKey == null || message == null || signature == null) {
        throw InvalidKeyException("Values must not be null")
      }

      return signingKey.verifySignature(message, signature)
    }

    @JvmStatic
    @Throws(InvalidKeyException::class)
    fun calculateSignature(signingKey: ECPrivateKey?, message: ByteArray?): ByteArray {
      if (signingKey == null || message == null) {
        throw InvalidKeyException("Values must not be null")
      }

      return signingKey.calculateSignature(message)
    }
  }
}
